sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala

```diff
@@ -101,8 +101,9 @@ private[this] object JsonPathParser extends RegexParsers {
 
 private[this] object SharedFactory {
   val jsonFactory = new JsonFactoryBuilder()
-    // Enabled for Hive compatibility
+    // The two options below enabled for Hive compatibility
     .enable(JsonReadFeature.ALLOW_UNESCAPED_CONTROL_CHARS)
+    .enable(JsonReadFeature.ALLOW_SINGLE_QUOTES)
```

Comment:
We also have the other Json built-in functions, but only these two functions behave differently from the others. It looks weird especially when Hive compatibility is not important for us.

Reply:
But the main reason why these were added is for Hive compatibility IIRC. It's a minimised change to keep the compatibility. Also, I guess these are all instances for Spark JSON functions corresponding to Hive?

Reply:
I am afraid that end users expect all our JSON functions have the same behavior. That means, either all the functions support SINGLE QUOTE or none of them supports it. Make the behavior configurable? For reading the external JSON data sources, users can control it https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala#L141-L149 

BTW, will this change cause the query result difference?

Reply:
> I am afraid that end users expect all our JSON functions have the same behavior. That means, either all the functions support SINGLE QUOTE or none of them supports it. Make the behavior configurable? For reading the external JSON data sources, users can control it

If we make it configurable, could we make all the behaivours configurable by using `JSONOptions` here as max suggested in https://github.com/apache/spark/pull/26965/files#r360562916 ? Some users might expect the same behaivours between json functions and json external reading, too.

> BTW, will this change cause the query result difference?

Ah, I think yes; after this PR, some query results have changed from `null` to a correct answer. We need to update the migration guide for 3.0.

Reply:
It's too corner case to document .. who will rely on the behaviour of `null` from a single quote JSON ... Users from Hive might depend on the single quote behaviours but It's pretty unlikely for Spark users to depend on single quoted JSONs.

What about we think about fixing it when it's actually requested from users?

Reply:
OK. That sounds OK to me. Let us wait for user requests.